tiledb/api/src/group/mod.rs

```diff
@@ -0,0 +1,745 @@
+        let c_context = self.context.capi();
+        let mut num: u64 = out_ptr!();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_get_metadata_num(
+                c_context,
+                Self::capi(self),
+                &mut num,
+            )
+        })?;
+        Ok(num)
+    }
+
+    pub fn metadata(&self, key: LookupKey) -> TileDBResult<Metadata> {
+        let c_context = self.context.capi();
+        let mut vec_size: u32 = out_ptr!();
+        let mut c_datatype: ffi::tiledb_datatype_t = out_ptr!();
+        let mut vec_ptr: *const std::ffi::c_void = out_ptr!();
+
+        let name: String = match key {
+            LookupKey::Index(index) => {
+                let mut key_ptr: *const std::ffi::c_char = out_ptr!();
+                let mut key_len: u32 = out_ptr!();
+                self.context.capi_return(unsafe {
+                    ffi::tiledb_group_get_metadata_from_index(
+                        c_context,
+                        Self::capi(self),
+                        index.try_into().unwrap(),
+                        &mut key_ptr,
+                        &mut key_len,
+                        &mut c_datatype,
+                        &mut vec_size,
+                        &mut vec_ptr,
+                    )
+                })?;
+                let c_key = unsafe { std::ffi::CStr::from_ptr(key_ptr) };
+                Ok(c_key.to_string_lossy().into_owned()) as TileDBResult<String>
+            }
+            LookupKey::Name(name) => {
+                let c_name = cstring!(name.as_ref() as &str);
+                self.context.capi_return(unsafe {
+                    ffi::tiledb_group_get_metadata(
+                        c_context,
+                        Self::capi(self),
+                        c_name.as_ptr(),
+                        &mut c_datatype,
+                        &mut vec_size,
+                        &mut vec_ptr,
+                    )
+                })?;
+                Ok(name.to_owned())
+            }
+        }?;
+        let datatype = Datatype::try_from(c_datatype)?;
+        Ok(Metadata::new_raw(name, datatype, vec_ptr, vec_size))
+    }
+
+    pub fn has_metadata_key<S>(&self, name: S) -> TileDBResult<Option<Datatype>>
+    where
+        S: AsRef<str>,
+    {
+        let c_context = self.context.capi();
+        let c_name = cstring!(name.as_ref());
+        let mut c_datatype: ffi::tiledb_datatype_t = out_ptr!();
+        let mut exists: i32 = out_ptr!();
+        self.context.capi_return(unsafe {
+            ffi::tiledb_group_has_metadata_key(
+                c_context,
+                Self::capi(self),
+                c_name.as_ptr(),
+                &mut c_datatype,
+                &mut exists,
+            )
+        })?;
+        if exists == 0 {
+            return Ok(None);
+        }
+
+        let datatype = Datatype::try_from(c_datatype)?;
+        Ok(Some(datatype))
+    }
+
+    pub fn consolidate_metadata<S>(
+        &self,
```

Comment:
This C API doc doesn't mention anything about the group open mode but this sure sounds like a mutating function call...

Anyway, as I've been mulling over the whole "some methods are `&mut self` and some aren't" thing, it's kind of occurred to me that `&mut self` to represent a write mode method doesn't really work because the API allows you to just get another handle, and then you have ownership of that second handle and can call whatever you want anyway. And this can go in the direction of "just let the C API give back the error and don't worry about it" or in the direction of "embed the permissions to call certain APIs in the type system".  If we had infinite time I'd want the latter but let's move ahead with the former for now.